include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+              x_k += object->vertex(i) *
+                     GeometryInfo<structdim>::d_linear_shape_function (xi, i);
+
+            if (delta_xi.norm() < 1e-7)
+              break;
+          }
+        while (true);
+
+        return x_k;
+      }
+    }
+  }
+
+
+
+  template <typename Iterator>
+  Point<Iterator::AccessorType::space_dimension>
+  project_to_object(const Iterator &object,
+                    const Point<Iterator::AccessorType::space_dimension> &trial_point)
+  {
+    const int spacedim = Iterator::AccessorType::space_dimension;
+    const int structdim = Iterator::AccessorType::structure_dimension;
+
+    Point<spacedim> projected_point = trial_point;
+
+    if (spacedim <= structdim)
+      return projected_point;
+    else if (structdim == 1 || structdim == 2)
+      {
+        using namespace internal::ProjectToObject;
+        // Try to use the special flat algorithm for quads (this is better
+        // than the general algorithm in 3D). This does not take into account
+        // whether projected_point is outside the quad, but we optimize along
+        // lines below anyway:
+        const int dim = Iterator::AccessorType::dimension;
+        const Manifold<dim, spacedim> &manifold = object->get_manifold();
+        if (structdim == 2 &&
+            dynamic_cast<const FlatManifold<dim,spacedim> *>(&manifold)
+            != nullptr)
+          {
+            projected_point = project_to_flat_object<Iterator, spacedim, structdim>(object, trial_point);
+          }
+        else
+          {
+            // We want to find a point on the convex hull (defined by the
+            // vertices of the object and the manifold description) that is
+            // relatively close to the trial point. This has a few issues:
+            //
+            // 1. For a general convex hull we are not guaranteed that a unique
+            //    minimum exists.
+            // 2. The independent variables in the optimization process are the
+            //    weights given to Manifold::get_new_point, which must sum to 1,
+            //    so we cannot use standard finite differences to approximate a
+            //    gradient.
+            //
+            // There is not much we can do about 1., but for 2. we can derive
+            // finite difference stencils that work on a simplex and rewrite
+            // the optimization problem to use those instead. Consider the
+            // structdim 2 case and let
+            //
+            // F(c0, c1, c2, c3) = Manifold::get_new_point(vertices, {c0, c1, c2, c3})
+            //
+            // then we seek to minimize the Euclidean distance between F(...)
+            // and trial_point. We can solve for c3 in terms of the other
+            // weights and get, for one coordinate direction
+            //
+            // d/dc0 ((x0 - F(c0, c1, c2, 1 - c0 - c1 - c2)))
+            //      = -2(x0 - F(...)) (d/dc0 F(...) - d/dc3 F(...))
```

Comment:
are you missing a square in the first line?

Also, would it be worthwhile making a distinction between the upright `d` on the lhs of the equation, and the partial derivatives on the rhs?

Reply:
I am missing a square. I'll fix the notation.

Reply:
I'd prefer to be lazy and keep 'd' so that we don't embed unicode characters for no reason. If you feel strongly about this I can change it (the result is prettier but I prefer to keep things to ascii if possible).

Reply:
I think the square is still missing.

I wasn't advocating for unicode, but maybe \partial? I think it's relevant to use this here because otherwise you would have `d/dc0 F` on *both* the left and right hand side, but on the left you mean something different from the right (namely, total vs partial derivative).